### PR TITLE
Update `--ext=rust` to support compiling the native extension from source

### LINKED ISSUE #7693: bundle gem --ext=rust does not install well as a gem out of the box
Hi.

### Describe the problem as clearly as you can

I've been trying to scaffold mixed rust/ruby code with magnus. So I tried the `bundle gem --ext=rust` command. And tweaked it a bit to make it work, but very minimally. Here is the result:

https://github.com/gl-yziquel/skeleton_rust_gem.git

Tagged: https://github.com/gl-yziquel/skeleton_rust_gem/releases/tag/gh-rubygems-7693

I added a Justfile to automate some tasks.

Bottom line: `just build` works fine and creates a gem in `pkg/`. However, `just bug-1` and `just bug-2` show that the scaffolded code is not designed, out of the box, to get the gem installed. I believe this is a shortcoming of the current scaffolding of rust gems.

```console
mini-me@virtucon ~/h/s/skeleton_rust_gem (master)> bundle exec rake install
cd tmp/x86_64-linux-gnu/skeleton_rust_gem/3.1.2
/usr/bin/ruby3.1 -I. ../../../../ext/skeleton_rust_gem/extconf.rb
checking for x86_64-linux-gnu-gcc... yes
checking for x86_64-linux-gnu-g++... yes
checking for x86_64-linux-gnu-gcc-ar... yes
cd -
cd tmp/x86_64-linux-gnu/skeleton_rust_gem/3.1.2
/usr/bin/gmake
generating /home/mini-me/home/stove/skeleton_rust_gem/target/release/libskeleton_rust_gem.so (release)
cargo rustc  --manifest-path /home/mini-me/home/stove/skeleton_rust_gem/ext/skeleton_rust_gem/Cargo.toml --target-dir /home/mini-me/home/stove/skeleton_rust_gem/target --lib --profile release -- -C linker=x86_64-linux-gnu-gcc -L native=/usr/lib -C link-arg=-Wl,-Bsymbolic-functions -C link-arg=-Wl,-z,relro -C link-arg=-Wl,-z,now -C link-arg=-lm
    Finished release [optimized] target(s) in 0.09s
cd -
/usr/bin/gmake install sitearchdir=../../../../lib/skeleton_rust_gem sitelibdir=../../../../lib/skeleton_rust_gem target_prefix=
generating /home/mini-me/home/stove/skeleton_rust_gem/target/release/libskeleton_rust_gem.so (release)
cargo rustc  --manifest-path /home/mini-me/home/stove/skeleton_rust_gem/ext/skeleton_rust_gem/Cargo.toml --target-dir /home/mini-me/home/stove/skeleton_rust_gem/target --lib --profile release -- -C linker=x86_64-linux-gnu-gcc -L native=/usr/lib -C link-arg=-Wl,-Bsymbolic-functions -C link-arg=-Wl,-z,relro -C link-arg=-Wl,-z,now -C link-arg=-lm
    Finished release [optimized] target(s) in 0.09s
installing skeleton_rust_gem.so to ../../../../lib/skeleton_rust_gem
/usr/bin/install -c -m 0755 skeleton_rust_gem.so ../../../../lib/skeleton_rust_gem
cp tmp/x86_64-linux-gnu/skeleton_rust_gem/3.1.2/skeleton_rust_gem.so tmp/x86_64-linux-gnu/stage/lib/skeleton_rust_gem/skeleton_rust_gem.so
skeleton_rust_gem 0.1.0 built to pkg/skeleton_rust_gem-0.1.0.gem.
rake aborted!
Running `gem install /home/mini-me/home/stove/skeleton_rust_gem/pkg/skeleton_rust_gem-0.1.0.gem` failed with the following output:

ERROR:  Error installing /home/mini-me/home/stove/skeleton_rust_gem/pkg/skeleton_rust_gem-0.1.0.gem:
        ERROR: Failed to build gem native extension.

    current directory: /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/ext/skeleton_rust_gem
cargo rustc --target-dir /home/mini-me/.local/share/gem/ruby/3.1.0/extensions/x86_64-linux/3.1.0/skeleton_rust_gem-0.1.0 --manifest-path /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/ext/skeleton_rust_gem/Cargo.toml --lib --release --locked -- -C linker\=x86_64-linux-gnu-gcc -L native\=/usr/lib -C link_arg\=-Wl,-Bsymbolic-functions -C link_arg\=-Wl,-z,relro -C link_arg\=-Wl,-z,now -l m -C debuginfo\=1
    Updating crates.io index
error: the lock file /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --locked flag and use --offline instead.

cargo failed, exit code 101

Gem files will remain installed in /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0 for inspection.
Results logged to /home/mini-me/.local/share/gem/ruby/3.1.0/extensions/x86_64-linux/3.1.0/skeleton_rust_gem-0.1.0/gem_make.out

/home/mini-me/.local/share/gem/ruby/3.1.0/gems/bundler-2.5.6/lib/bundler/gem_helper.rb:210:in `sh'
/home/mini-me/.local/share/gem/ruby/3.1.0/gems/bundler-2.5.6/lib/bundler/gem_helper.rb:101:in `install_gem'
/home/mini-me/.local/share/gem/ruby/3.1.0/gems/bundler-2.5.6/lib/bundler/gem_helper.rb:57:in `block in install'
/home/mini-me/.local/share/gem/ruby/3.1.0/gems/rake-13.2.1/exe/rake:27:in `<top (required)>'
Tasks: TOP => install
(See full trace by running task with --trace)
mini-me@virtucon ~/h/s/skeleton_rust_gem (master) [1]> cd pkg
mini-me@virtucon ~/h/s/s/pkg (master)> gem install skeleton_rust_gem-0.1.0.gem
Building native extensions. This could take a while...
ERROR:  Error installing skeleton_rust_gem-0.1.0.gem:
        ERROR: Failed to build gem native extension.

    current directory: /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/ext/skeleton_rust_gem
cargo rustc --target-dir /home/mini-me/.local/share/gem/ruby/3.1.0/extensions/x86_64-linux/3.1.0/skeleton_rust_gem-0.1.0 --manifest-path /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/ext/skeleton_rust_gem/Cargo.toml --lib --release --locked -- -C linker\=x86_64-linux-gnu-gcc -L native\=/usr/lib -C link_arg\=-Wl,-Bsymbolic-functions -C link_arg\=-Wl,-z,relro -C link_arg\=-Wl,-z,now -l m -C debuginfo\=1
    Updating crates.io index
error: the lock file /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0/Cargo.lock needs to be updated but --locked was passed to prevent this
If you want to try to generate the lock file without accessing the network, remove the --locked flag and use --offline instead.

cargo failed, exit code 101

Gem files will remain installed in /home/mini-me/.local/share/gem/ruby/3.1.0/gems/skeleton_rust_gem-0.1.0 for inspection.
Results logged to /home/mini-me/.local/share/gem/ruby/3.1.0/extensions/x86_64-linux/3.1.0/skeleton_rust_gem-0.1.0/gem_make.out
mini-me@virtucon ~/h/s/s/pkg (master) [1]>
```

Ideally, I'd expect the scaffolding of such a rust extension gem to be installable in some way. Currently, it does not seem to be the case.

### Did you try upgrading rubygems & bundler?

No. not yet. On my todo list.

### PR DESCRIPTION
**Pull request by karreiro**
## What was the end-user or developer problem that led to this PR?

I needed to apply some tiny changes to the skeleton generated by the `--ext=rust` option to compile my native extension from source.

## What is your fix for the problem, implemented in this PR?

When gem consumers install a gem with a native extension and their platform is not included the pre-compiled bundles, they need to compile it locally and rely on the `extconf.rb` file.

Referencing the `extconf.rb` file in the gemspec is required  to make that work. We also need to include `rb_sys` as a dependency because the Rust `extconf.rb` depends on it.

Here's a demo of this change in action :)

https://github.com/rubygems/rubygems/assets/1079279/54e543a8-0827-4fce-bab6-1bb8c0740b42

Closes #7693.

### CODE REVIEW

**Comment by karreiro** [comment]
👋 Hey @simi, thanks for chiming in on this PR. I've noticed the branch was outdated with the `master`, so I just updated it.

I'd appreciate your review when you have a moment :)

**Comment by simi** [comment]
Hello @karreiro and thanks for your PR. Can you please explain how to reproduce your errors? I just tried in latest bundler to create empty skeleton gem, pack it into gem and install with no problems. I have did following.

```bash
$ gem list | grep rb_sys  # ensure rb_sys is not installed already
$ bundle gem --ext=rust ryba > /dev/null # create empty gem
$ cd ryba # move into gem directory
$ cargo c # resolve the Cargo.lock somehow
$ git add Cargo.lock # add lock to the final gem, since gem files are collected from git by default
$ gem build --force # build gem, ignore TODO in gemspec
$ gem install ryba-0.1.0.gem # install local gem
```

**Comment by karreiro** [comment]
Hello @simi,

Thank **you** for reviewing this PR :)

## Setup

To reproduce the error, let's create 3 ryba gems:

### ryba1
```
- Run `bundle gem --ext=rust ryba1 > /dev/null`
- Run `cd ryba1`
- Run `cargo c`
- Run `git add Cargo.lock`
- Run `gem build --force`
```

### ryba2
```
- Run `bundle gem --ext=rust ryba2 > /dev/null`
- Run `cd ryba2`
- Run `cargo c`
- Run `git add Cargo.lock`
- Update the `ryba2.gemspec` file to use `spec.extensions = ["ext/ryba2/extconf.rb"]` instead of `spec.extensions = ["ext/ryba2/Cargo.toml"]`
- Run `gem build --force`
```

### ryba3
```
- Run `bundle gem --ext=rust ryba3 > /dev/null`
- Run `cd ryba3`
- Run `cargo c`
- Run `git add Cargo.lock`
- Update the `ryba3.gemspec` file to use `spec.extensions = ["ext/ryba3/extconf.rb"]` instead of `spec.extensions = ["ext/ryba3/Cargo.toml"]`
- Update the `ryba3.gemspec` file to include `spec.add_dependency "rb_sys", "~> 0.9.91"`
- Run `gem build --force`
```


## Results

Now, we have 3 .gem files:

- `ryba1-0.1.0.gem`
- `ryba2-0.1.0.gem`
- `ryba3-0.1.0.gem`

Let's move to a machine without Rust installed now.

### ryba1 (result)

When we run `gem install ryba1-0.1.0.gem`, we get an error like this:

```
Building native extensions. This could take a while...
ERROR:  Error installing ryba1-0.1.0.gem:
        ERROR: Failed to build gem native extension.

    cargo metadata failed No such file or directory - cargo

Gem files will remain installed in /opt/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/gems/ryba-0.1.0 for inspection.
Results logged to /opt/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/extensions/x86_64-linux/3.1.0/ryba-0.1.0/gem_make.out
```

### ryba2 (result)

When we run `gem install ryba2-0.1.0.gem`, we get an error like this:

```
Building native extensions. This could take a while...
ERROR:  Error installing ryba2-0.1.0.gem:
        ERROR: Failed to build gem native extension.

    current directory: /opt/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/gems/ryba2-0.1.0/ext/ryba2
/opt/rubies/ruby-3.1.4/bin/ruby -I /opt/rubies/ruby-3.1.4/lib/ruby/site_ruby/3.1.0 extconf.rb
*** extconf.rb failed ***
Could not create Makefile due to some reason, probably lack of necessary
libraries and/or headers.  Check the mkmf.log file for more details.  You may
need configuration options.

Provided configuration options:
        --with-opt-dir
        --without-opt-dir
        --with-opt-include
        --without-opt-include=${opt-dir}/include
        --with-opt-lib
        --without-opt-lib=${opt-dir}/lib
        --with-make-prog
        --without-make-prog
        --srcdir=.
        --curdir
        --ruby=/opt/rubies/ruby-3.1.4/bin/$(RUBY_BASE_NAME)
<internal:/opt/rubies/ruby-3.1.4/lib/ruby/site_ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:88:in `require': cannot load such file -- rb_sys/mkmf (LoadError)
        from <internal:/opt/rubies/ruby-3.1.4/lib/ruby/site_ruby/3.1.0/rubygems/core_ext/kernel_require.rb>:88:in `require'
        from extconf.rb:4:in `<main>'

extconf failed, exit code 1

Gem files will remain installed in /opt/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/gems/ryba2-0.1.0 for inspection.
Results logged to /opt/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/extensions/x86_64-linux/3.1.0/ryba2-0.1.0/gem_make.out
```

### ryba3 (result)

When we run `gem install ryba3-0.1.0.gem`, it's going to work 🎉

**Comment by karreiro** [comment]
I just rebased this branch :)

**Comment by deivid-rodriguez** [comment]
Is this fixing the same problem as #7695, closing #7693? I'm surprised that RubyGems will automatically install `rust` if not present when finding a rust extension.

Not sure how I feel about that. I guess it's handy although I'd never expect `gem install` to install a programming language... 😅

**Comment by karreiro** [comment]
> Is this fixing the same problem as https://github.com/rubygems/rubygems/pull/7695, closing https://github.com/rubygems/rubygems/issues/7693? I'm surprised that RubyGems will automatically install rust if not present when finding a rust extension.

Yes, it addresses the same issue as https://github.com/rubygems/rubygems/pull/7695 and closes https://github.com/rubygems/rubygems/issues/7693.

However, this PR goes a bit further by moving `rb_sys` to the gemspec, as we alreqady have this line in the [`extconf-rust.rb.tt`](https://github.com/rubygems/rubygems/blob/3fdc2f627e6bb476b722c7d37b61074430dea82e/bundler/lib/bundler/templates/newgem/ext/newgem/extconf-rust.rb.tt#L4) file:

```ruby
require "rb_sys/mkmf"
```

> Not sure how I feel about that. I guess it's handy although I'd never expect gem install to install a programming language... 😅

heh, I share the same feeling... the `extconf.rb` file is quite powerful 😅

We could possibly propose making that opt-in on `rb_sys`. Still, the main goal of this PR is to ensure that the gem's scaffold works end-to-end with all dependencies in place, so new developers may successfully compile their gems from source with the `extconf.rb` file.

**Comment by deivid-rodriguez** [comment]
> However, this PR goes a bit further by moving rb_sys to the gemspec, as we alreqady have this line in the [extconf-rust.rb.tt](https://github.com/rubygems/rubygems/blob/3fdc2f627e6bb476b722c7d37b61074430dea82e/bundler/lib/bundler/templates/newgem/ext/newgem/extconf-rust.rb.tt#L4) file.

Cool, yeah, that makes sense. If `rb_sys` is required by the `extconf.rb` file, and that file is run at installation time, then `rb_sys` should be a runtime dependency 👍.

> We could possibly propose making that opt-in on rb_sys. Still, the main goal of this PR is to ensure that the gem's scaffold works end-to-end with all dependencies in place, so new developers may successfully compile their gems from source with the extconf.rb file.

Yep, seems something separate and should not affect this PR. Still surprising though!

**Review by deivid-rodriguez — APPROVED** [review]
This looks good to me, but I'll let @simi give final approval!

**Comment by deivid-rodriguez** [comment]
@simi Does this look good to you? I think it makes sense and it has been proposed independently 3 times (here, #7695 and #7608).

**Comment by simi** [comment]
@deivid-rodriguez yes, I'll test and merge.

**Comment by deivid-rodriguez** [comment]
@simi I did test this myself this month and it worked fine, I think it'd be fine to merge if you don't have too much time to do the testing yourself.

@ianks Since you proposed a slightly overlapping PR at #7608 (and you're rb-sys author after all :)), would like to give this a quick look? We can them move on to #7608 after merging this.

**Comment by simi** [comment]
@deivid-rodriguez I'll test soon :pray:, initially looks good.

**Comment by deivid-rodriguez** [comment]
Ok, thank you @simi 😃.

**Review by ianks — APPROVED** [review]

**Review by simi — APPROVED** [review]

**Pull request merged**
